### PR TITLE
Implementacion del card Home page: Frontend - News

### PR DESCRIPTION
**Pull request by FloresFelix**
Summary
implementación del card Home page: Frontend - News

Screenshot

Trello card
https://trello.com/c/xXBJJpOl

### CODE REVIEW

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import ar.com.wolox.android.R
+import ar.com.wolox.android.example.model.ItemNewsModel
+import com.bumptech.glide.Glide
+import org.ocpsoft.prettytime.PrettyTime
+import java.text.SimpleDateFormat
+import java.util.Date
+import kotlin.collections.ArrayList
+
+class NewsAdapter constructor(var context: Context, var listNews: ArrayList<ItemNewsModel>) : RecyclerView.Adapter<NewsAdapter.NewsViewHolder>() {
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt]: Never forget run the "Code -> Reformat Code" Android Studio tool to make sure that your code is well indented.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 2]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
     override fun layout() = R.layout.fragment_news
 
     override fun init() {
+        presenter.setListNews()
+    }
+
+    override fun setListeners() {
+        binding.newsSwipeToRefresh.setOnRefreshListener {
+            presenter.refreshNews()
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: Rename this presenter function to onSwipeRefresh(), remember that the fragment notify to presenter asking what should I do? Usually, the fragment don't take decision by itself.

File: app/src/main/java/ar/com/wolox/android/example/model/ItemNewsModel.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package ar.com.wolox.android.example.model
+
+data class ItemNewsModel constructor(
+    val newPhoto: Int,
+    val newsTitle: String,
+    val newsContent: String,
+    val newsTime: String,
+    val newsLike: Int
+)
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/model/ItemNewsModel.kt:9]: You will need to change this names when you perform the integration with backend.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_news, parent, false)
+        return NewsViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
+
+        holder.newsContent.setText(listNews[position].newsContent)
+        holder.newsTitle.setText(listNews[position].newsTitle)
+        holder.newsLike.setImageResource(listNews[position].newsLike)
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt]: `holder.apply {} `

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        Glide.with(context).load(R.mipmap.ic_launcher).into(holder.newsPhoto)
+        val time = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(listNews[position].newsTime).time
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsAdapter.kt]: move the date format to a constant

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 6]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
...
+
+    override fun loadRecyclerView(listNews: ArrayList<ItemNewsModel>) {
+        binding.newsRecyclerView.layoutManager = LinearLayoutManager(requireContext())
+        binding.newsRecyclerView.adapter = NewsAdapter(requireContext(), listNews)
+    }
+
+    override fun showEmptyNews() {
+        binding.imageViewEmptyNews.visibility = View.VISIBLE
+        binding.textViewEmptyNews.visibility = View.VISIBLE
+    }
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: `with(binding)`

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 7]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
...
+
+    override fun setListeners() {
+        binding.newsSwipeToRefresh.setOnRefreshListener {
+            presenter.refreshNews()
+        }
+    }
+
+    override fun loadRecyclerView(listNews: ArrayList<ItemNewsModel>) {
+        binding.newsRecyclerView.layoutManager = LinearLayoutManager(requireContext())
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: Use "with" function like I mentioned above. If don't understand ask us, no problem.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt [hunk 8]
```diff
@@ -1,6 +1,25 @@
...
 import javax.inject.Inject
 
-class NewsPresenter @Inject constructor() : BasePresenter<NewsView>()
\ No newline at end of file
+class NewsPresenter @Inject constructor() : BasePresenter<NewsView>() {
+
+    fun setListNews() {
+        var list = ArrayList<ItemNewsModel>()
+        for (num in 1..20) {
+            if (num % 2 == 0) {
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt:17]: Why do you need this? 🤔

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 9]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
...
+        binding.newsRecyclerView.adapter = NewsAdapter(requireContext(), listNews)
+    }
+
+    override fun showEmptyNews() {
+        binding.imageViewEmptyNews.visibility = View.VISIBLE
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: Same as above

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt [hunk 10]
```diff
@@ -1,6 +1,25 @@
...
+                list.add(ItemNewsModel(R.drawable.ic_profile_icon, "Ali Connors", "I´ll be in you neighborhood doing errands...", "2021-07-29T11:00:00.000Z", R.drawable.ic_favorite))
+            } else {
+                list.add(ItemNewsModel(R.drawable.ic_profile_icon, "Ali Connors", "I´ll be in you neighborhood doing errands...", "2021-07-29T09:00:00.000Z", R.drawable.ic_favorite_border))
+            }
+        }
+        if (list.isEmpty()) view?.showEmptyNews()
+        view?.loadRecyclerView(list)
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt]: You should do this in an if/else statement. `if (list.isEmpty) showEmptyNews else showNews`

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 11]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
     override fun layout() = R.layout.fragment_news
 
     override fun init() {
+        presenter.setListNews()
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: You should rename this presenter function to onInit and make private the setListNews function, so, call setListnews function from onInit. Remember, the fragment allways ask what should make?

**Review by ffasitella — CHANGES_REQUESTED** [review]
Left some comments. Check this PR, it has some merge conflicts. You shoul merge master to this branch to solve them.
Also, add a screenshot to show us how this looks 👀

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt (hunk 10 above)
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt]: Rename this function to any more descriptive, the presenter don't know about Recycler Views. You could rename to loadNewsList.

File: app/src/main/res/drawable/ic_add.xml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Comment (jcruz-wolox) [inline app/src/main/res/drawable/ic_add.xml]: Move to dimens.xml file the width and height values.

File: app/src/main/res/drawable/ic_favorite.xml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Comment (jcruz-wolox) [inline app/src/main/res/drawable/ic_favorite.xml]: Same as above, move to dimens.xml width and height values.

File: app/src/main/res/drawable/ic_favorite_border.xml [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Comment (jcruz-wolox) [inline app/src/main/res/drawable/ic_favorite_border.xml]: Same as above.

File: app/src/main/res/drawable/ic_news_icons.xml [hunk 15]
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Comment (jcruz-wolox) [inline app/src/main/res/drawable/ic_news_icons.xml]: Same as above.

File: app/src/main/res/layout/fragment_news.xml [hunk 16]
```diff
@@ -1,14 +1,57 @@
...
+            android:id="@+id/textViewEmptyNews"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/empty_news"
+            android:visibility="gone"
+            app:layout_constraintEnd_toEndOf="@+id/imageViewEmptyNews"
+            app:layout_constraintStart_toStartOf="@+id/imageViewEmptyNews"
+            app:layout_constraintTop_toBottomOf="@id/imageViewEmptyNews" />
 
+
```
Comment (jcruz-wolox) [inline app/src/main/res/layout/fragment_news.xml]: Remove empty lines.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt [hunk 17]
```diff
@@ -1,6 +1,25 @@
...
+    }
+    fun refreshNews() {
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsPresenter.kt]: You could rename to onRefreshNews and call setListNews() function

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt [hunk 18]
```diff
@@ -10,5 +13,26 @@ class NewsFragment @Inject constructor() : WolmoFragment<FragmentNewsBinding, Ne
...
+        binding.newsRecyclerView.layoutManager = LinearLayoutManager(requireContext())
+        binding.newsRecyclerView.adapter = NewsAdapter(requireContext(), listNews)
+    }
+
+    override fun showEmptyNews() {
+        binding.imageViewEmptyNews.visibility = View.VISIBLE
+        binding.textViewEmptyNews.visibility = View.VISIBLE
+    }
+    override fun refreshView() {
+        presenter.setListNews()
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: Call to onRefreshNews() function. Below I explain how

**Review by jcruz-wolox — CHANGES_REQUESTED** [review]
Hello Félix!! 

I have any comments.

One suggestion, you can make comments from here (Github PR), i think that you is making comments by responding the github notifications emails, shouldn't response them because it generates spam in the PR.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt (hunk 11 above)
Comment (FloresFelix) [reply app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: sorry I don't understand...

Comment (jcruz-wolox) [reply app/src/main/java/ar/com/wolox/android/example/ui/viewpager/news/NewsFragment.kt]: Sorry, I confused you. The idea is that you convert to private setListNews and create onInit function to call it here. Example:

NewsFragment.kt

`override fun init() {
    presenter.onInit()
}`

-------------------------------

Presenter:

```
// New function called onInit()
fun onInit() {
    setListNews()
}
```

```
// the setListNews function converted to private. you'll be able to call this function also when you refresh the news list
private fun setListNews() {
 
}
```

**Review by ffasitella — APPROVED** [review]
Good job! In a real project, you will fight merge conflicts very often. Well done 👍🏽 
Let's wait for Jona's approval!

**Review by jcruz-wolox — APPROVED** [review]

**Pull request merged**